pset/blinder.go

```diff
@@ -0,0 +1,314 @@
+)
+
+type blinder struct {
+	pset             *Pset
+	blindingPrivkeys [][]byte
+	blindingPubkeys  [][]byte
+}
+
+// NewBlinder returns a new instance of blinder, if the passed Pset struct is
+// in a valid form, else an error.
+func NewBlinder(pset *Pset, blindingPrivkeys, blindingPubkeys [][]byte) (
+	*blinder,
+	error,
+) {
+	if err := pset.SanityCheck(); err != nil {
+		return nil, err
+	}
+
+	return &blinder{
+		pset:             pset,
+		blindingPrivkeys: blindingPrivkeys,
+		blindingPubkeys:  blindingPubkeys,
+	}, nil
+}
+
+//BlindOutputs method blinds pset's output
+func (b *blinder) BlindOutputs() error {
+	err := b.validate()
+	if err != nil {
+		return err
+	}
+
+	unblindOutputs, err := b.unblindInputs()
+	if err != nil {
+		return err
+	}
+
+	outputValues := make([]uint64, 0)
+	//TODO check if looping over UnsignedTx's Outputs is correct?
+	for _, output := range b.pset.UnsignedTx.Outputs {
+		var val [9]byte
+		copy(val[:], output.Value)
+		value, err := confidential.ElementsToSatoshiValue(val)
+		if err != nil {
+			return err
+		}
+		outputValues = append(outputValues, value)
+	}
+
+	inputAbfs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputAbfs = append(inputAbfs, v.AssetBlindingFactor)
+	}
+
+	inputVbfs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputVbfs = append(inputVbfs, v.ValueBlindingFactor)
+	}
+
+	inputAgs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputAgs = append(inputAgs, v.Asset)
+	}
+
+	inputValues := make([]uint64, 0)
+	for _, v := range unblindOutputs {
+		inputValues = append(inputValues, v.Value)
+	}
+
+	outputVbfs, outputAbfs, err := b.generateOutputBlindingFactors(
+		inputValues,
+		outputValues,
+		inputAbfs,
+		inputVbfs,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = b.blindOutputs(
+		outputValues,
+		outputAbfs,
+		outputVbfs,
+		inputAgs,
+		inputAbfs,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *blinder) validate() error {
+	for _, input := range b.pset.Inputs {
+		if input.NonWitnessUtxo == nil && input.WitnessUtxo == nil {
+			return errors.New("all inputs must contain a non witness " +
+				"utxo or a witness utxo")
+		}
+	}
+
+	if len(b.blindingPrivkeys) != len(b.pset.Inputs) {
+		return errors.New("blinding private keys do not match the number" +
+			" of inputs")
+	}
+
+	if len(b.blindingPubkeys) != (len(b.pset.Outputs) - 1) {
+		return errors.New("blinding public keys do not match the number " +
+			"of outputs (fee excluded)")
+	}
+	return nil
+}
+
+func (b *blinder) unblindInputs() ([]confidential.UnblindOutputResult, error) {
+	ctx, _ := secp256k1.ContextCreate(secp256k1.ContextBoth)
+	defer secp256k1.ContextDestroy(ctx)
+	unblindOutputs := make([]confidential.UnblindOutputResult, 0)
+	for index, input := range b.pset.Inputs {
+		var prevout *transaction.TxOutput
+		if input.NonWitnessUtxo != nil {
+			//TODO: check if prevout should be created like below
+			prevout = input.NonWitnessUtxo.Outputs[index]
```

Comment:
here we should use the unsigned tx input's `Index` instead of the actual index of the input in the transaction.
This sounds like a joke, but the `TxHash` and `Index` of a tx input actually refer to the specif output at index `Index` of the transaction with id `TxHash`:
```suggestion
			prevout = input.NonWitnessUtxo.Outputs[input.Index]
```

Suggestion relies on iterating over unsigned tx's inputs

Reply:
Done